### PR TITLE
Fix handling of `white-space` property for wrapping

### LINKED ISSUE #30241: `white-space: nowrap` not respected for text
```html
<div style="width: 200px; border: solid;">
    <span style="white-space: nowrap; font-size: 50px">large font and wide</span>
</div>
```

Layout 2020:

image

Firefox:

image

### LINKED ISSUE #30242: Atomic inline elements do not wrap when the line has content from other nesting levels
```html
<div style="width: 200px; border: solid;">
    <div style="display: inline-block; width: 100px; background: teal;">hey</div>
    <span>
        <div style="display: inline-block; width: 150px; background: green;">hey</div>
    </span>
</div>
```
Layout 2020:
image

Firefox:
image

### PR DESCRIPTION
**Pull request by mrobinson**
- Properly handle the `white-space` property when it prevents wrapping in text runs.
 - Line breaking should be allowed for atomics even when previous line content was added in another inline box.

Fix #30241.
Fix #30242.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #30241 and #30242.
- [x] There are tests for these changes

### CODE REVIEW

**Comment by mrobinson** [comment]
@bors-servo try=wpt-2020

This depends on #30243, so it's currently marked as a draft.

**Comment by mrobinson** [comment]
Explanation for two new failures:
 - `white-space-normal-011.html`: this is failing because Servo doesn't properly handle the `ch` unit.
 - `white-space-wrap-after-nowrap-001.html`: This is failing because `nowrap` is now respected in text runs, but there is one more case that we fail to propagate it. I plan to fix this in a followup.

**Comment by mrobinson** [comment]
@Loirooriol Okay. This is ready to review. It turns out that the original order of operations for deciding on line breaking was important.

File: components/layout_2020/flow/inline.rs [hunk 1]
```diff
@@ -731,7 +772,17 @@ impl InlineFormattingContext {
                     &mut ifc.current_line.inline_position,
                     true, /* at_end_of_inline_element */
                 );
-                ifc.current_nesting_level = partial.parent_nesting_level
+
+                let had_content = ifc.current_nesting_level.has_content;
+                ifc.current_nesting_level = partial.parent_nesting_level;
+
+                // If the inline box that we just finished had any content at all, we want to propogate
```
Comment (Loirooriol) [inline components/layout_2020/flow/inline.rs:779]: Typo: propagate

**Review by Loirooriol — APPROVED** [review]

**Comment by mrobinson** [comment]
I have opened #30288 for the failure of  /_mozilla/css/whitespace_nowrap_line_breaking_a.html.

**Pull request merged**